patternfly/patternfly-elements
# feat: persist _debugLog in localStorage

## Persist _debugLog in localStorage instead of PFElement class object

This will help avoid the timing issues we currently have when enabling debugLog.  By reading it from localStorage, it can be set once and the page refreshed.  Debugging will remain enabled on that browser, for that domain, until it's turned off or cache is cleared.

### Related issues

- (#1550) the global PfeConfig object is another approach that could make debugLog easier to enable, though it wouldn't last across page refreshes.


### Preview


Link(s) to demo page(s) where this element can be viewed:
- [Link](https://deploy-preview-1570--patternfly-elements.netlify.app/)


### What has changed and why



- the existing PFElement.debugLog function now reads and writes `localStorage.pfeLog` instead of `PFElement._debugLog`.


### Testing instructions



- [x] *Test by setting localStorage manually*
  1. visit the pfe examples on localhost:1234 (or whatever origin yours launches on)
  2. open the dev console and type `localStorage.pfeLog = true` and press enter
  3. refresh the page and behold all the logs
  4. navigate to another element's demo page and see its logs too
- [x] *Test by manually running PFElement.debugLog(true)*
  1. visit the pfe examples on localhost:1234 (or whatever origin yours launches on)
  2. now you must get a reference to the PFElement class, so open the dev console and type `let {default:PFElement} = await import("../../pfelement/dist/pfelement.js");` and press enter
  3. now enable debug logging by typing `PFElement.debugLog(true)` and press enter
  4. refresh the page and behold all the logs
  5. navigate to another element's demo page and see its logs too
- [x] *Test by running PFElement.debugLog(true) in your element's code*
  1. in your local dev environment, add `PFElement.debugLog(true)` to an element of your choice.  put the code right before your element's class statement.
  2. now you must get a reference to the PFElement class, so open the dev console and type `window.PFElement = await import("path/to/pfelement.js")` and press enter
  4. after a build, view the demo page (or storybook, or wherever)
  5. navigate to another element's demo page and see its logs too (it's a global setting for all elements, not just your chosen element.)


#### Browser requirements

Your component should work in all of the following environments:



### Ready-for-merge Checklist





### Merging

Please **squash** when merging and ensure your commit message uses [conventional commit](https://www.conventionalcommits.org/en/v1.0.0/#summary) formatting.

**Be sure to share your updates with the [patternfly-elements-contribute@redhat.com](mailto:patternfly-elements-contribute@redhat.com) mailing list!**
